platform/ios/src/MGLMapView.mm

```diff
@@ -3187,6 +3187,14 @@ - (void)addAnnotations:(NS_ARRAY_OF(id <MGLAnnotation>) *)annotations
     {
         NSAssert([annotation conformsToProtocol:@protocol(MGLAnnotation)], @"annotation should conform to MGLAnnotation");
 
+        if ([annotation conformsToProtocol:@protocol(MGLFeature)])
+        {
+            static dispatch_once_t onceToken;
+            dispatch_once(&onceToken, ^{
+                NSLog(@"Annotations that conform to the MGLFeature protocol do not keep their `identifier` or `attributes` properties when added to the map as annotations. Add this feature using runtime styling if you intend to later query the map for it. This warning is only shown once. %@", annotation.description);
```

Comment:
I wonder if we could move this warning to when the developer attempts to query the map. Inside `-visibleFeaturesAt:inStyleLayersWithIdentifiers:`, warn if any of the values in `_annotationContextsByAnnotationTag` or keys in `_annotationTagsByAnnotation` conforms to MGLFeature. Or to make it more performant, set a flag here that enables the check later in `-visibleFeaturesAt:inStyleLayersWithIdentifiers:`.